lib/rules/no-unused-prop-types.js

```diff
@@ -725,6 +725,70 @@ module.exports = {
     }
 
     /**
+     * Marks all props found inside ObjectTypeAnnotaiton as declared.
+     * 
+     * Modifies the declaredProperties object
+     * @param {ASTNode} propTypes 
+     * @param {Object} declaredPropTypes 
+     * @returns {Boolean} True if propTypes should be ignored (e.g. when a type can't be resolved, when it is imported)
+     */
+    function declarePropTypesForObjectTypeAnnotation(propTypes, declaredPropTypes) {
+      let ignorePropsValidation = false;
+
+      iterateProperties(propTypes.properties, (key, value) => {
+        if (!value) {
+          ignorePropsValidation = true;
+          return;
+        }
+
+        let types = buildTypeAnnotationDeclarationTypes(value, key);
+        if (types === true) {
+          types = {};
+        }
+        types.fullName = key;
+        types.name = key;
+        types.node = value;
+        declaredPropTypes.push(types);
+      });
+
+      return ignorePropsValidation;
+    }
+
+    /**
+     * Marks all props found inside IntersectionTypeAnnotation as declared.
+     * Since InterSectionTypeAnnotations can be nested, this handles recursively. 
+     * 
+     * Modifies the declaredPropTypes object
+     * @param {ASTNode} propTypes 
+     * @param {Object} declaredPropTypes 
+     * @returns {Boolean} True if propTypes should be ignored (e.g. when a type can't be resolved, when it is imported)
+     */
+    function declarePropTypesForIntersectionTypeAnnotation(propTypes, declaredPropTypes) {
+      return propTypes.types.reduce((ignorePropsValidation, annotation) => {
+        // If we already decided to skip props validation then we don't have to continue processing anything else
+        if (ignorePropsValidation) {
+          return ignorePropsValidation;
+        }
+
+        if (annotation.type === 'ObjectTypeAnnotation') {
+          ignorePropsValidation = declarePropTypesForObjectTypeAnnotation(annotation, declaredPropTypes);
+        } else {
+          const typeNode = typeScope(annotation.id.name);
+
+          if (!typeNode) {
+            ignorePropsValidation = true;
+          } else if (typeNode.type === 'IntersectionTypeAnnotation') {
+            ignorePropsValidation = declarePropTypesForIntersectionTypeAnnotation(typeNode, declaredPropTypes);
+          } else {
+            ignorePropsValidation = declarePropTypesForObjectTypeAnnotation(typeNode, declaredPropTypes);
+          }
+        }
+
+        return ignorePropsValidation;
```

Comment:
i can't decide if this style, or early-returning each place this var is assigned, is better

Reply:
Honestly I'm in general not really happy about the fact that a function with a name `declarePropTypes***` returns a boolean at all. It's a bit confusing. That's why I liked this style a bit more, because it's very explicit to say what those methods return and what the boolean actually means.

Perhaps a `try/catch` may be even better?

Reply:
We could rename the function instead :-)